arm-hyp+aarch64 haskell: add VCPURegVMPIDR(_EL2)

lsf37 opened the pull request:
This brings the spec in line with seL4/seL4#1181.

Test with: seL4/seL4#1181

lsf37 commented:
Ok, this is now ready for review. The AARCH64 proof update has to be done separately, will add a PR to the aarch64 branch fro that later.

corlewis commented on proof/crefine/ARM_HYP/Retype_C.thy:
```diff
@@ -5574,7 +5574,7 @@ proof -
     apply (clarsimp simp: rf_sr_def cstate_relation_def carch_state_relation_def
                           cmachine_state_relation_def Let_def h_t_valid_clift_Some_iff)
-    apply (subgoal_tac "region_is_bytes p 456 \<sigma>'")
+    apply (subgoal_tac "region_is_bytes p 464 \<sigma>'")
```
It's hard to tell from the diff, but can you use `?vcpusz` here as well?

Xaphiosis replied:
Yeah, we're still in reach of the lemma statement, does it not work?

lsf37 replied:
Let me check, I thought this was already a different lemma. Would definitely be nicer to use here if not.

lsf37 replied:
You are absolutely right. Will update.

Xaphiosis commented on proof/crefine/ARM_HYP/Retype_C.thy:
```diff
@@ -5360,7 +5360,7 @@ proof -
   let ?zeros = "from_bytes (replicate (size_of TYPE(vcpu_C)) 0) :: vcpu_C"
 
-  have "size_of TYPE(vcpu_C) = 456" (is "_ = ?vcpusz")
+  have "size_of TYPE(vcpu_C) = ?vcpusz"
```
I'd add a comment that this is a sanity check, otherwise it's unused now and not declaring the var

lsf37 replied:
Good point, will add.

Xaphiosis commented on proof/crefine/ARM_HYP/Wellformed_C.thy:
```diff
@@ -43,7 +43,7 @@ abbreviation
   pd_Ptr :: "32 word \<Rightarrow> (pde_C[2048]) ptr" where "pd_Ptr == Ptr"
 abbreviation
-  regs_C_Ptr :: "addr \<Rightarrow> (machine_word_len word[42]) ptr" where "regs_C_Ptr \<equiv> Ptr"
+  regs_C_Ptr :: "addr \<Rightarrow> (machine_word_len word[43]) ptr" where"regs_C_Ptr \<equiv> Ptr"
```
I think I did something to improve this on AARCH64. Might be worth seeing what the lines you changed in this PR look like there in case there's improvement that we already figured out.

Xaphiosis replied:
ah yes, the `value_type vcpuregs_count = "length enum_vcpureg"` failing; so I tried to but did not succeed. any chance you could have a stab at that now?

lsf37 replied:
Hm, it probably needs some code generator setup and be done outside the Arch locale. I can have a look.

lsf37 replied:
It turns out there is a constant in C `seL4_VCPUReg_Num` for which we can declare `seL4_VCPUReg_Num_def[code]` and that then works fine.

lsf37 commented:
Have addressed the review feedback. Now that AARCH64 CRefine is merged, I can add the AARCH64 part of the proofs (and deploy the `seL4_VCPUReg_Num` value type there).
